lib/snapshot.go

```diff
@@ -1036,12 +1032,27 @@ func (snap *Snapshot) snapshotProcessBlockNoLock(blockNode *BlockNode) {
 
 		// Update the snapshot epoch metadata in the snapshot DB.
 		for ii := 0; ii < MetadataRetryCount; ii++ {
+			// We have to wait for the checksum threads to finish or else we'll set a stale checksum for
+			// our snapshot height.
+			glog.V(1).Infof("Snapshot.SnapshotProcessBlock: Waiting for checksum to finish updating...")
+			if localErr := snap.Checksum.Wait(); localErr != nil {
+				glog.Errorf("Snapshot.SetSnapshotChunk: Problem waiting for the checksum: %v", localErr)
+				time.Sleep(1 * time.Second)
+				continue
+			}
+			// For good measure, let's also wait for the operation queue to empty. This will ensure
+			// that any pending checksum operations are processed before we proceed.
+			glog.V(1).Infof("Snapshot.SnapshotProcessBlock: Waiting for all operations to " +
+				"clear so we can update snapshot checksum...")
+			snap.WaitForAllOperationsToFinish()
```

Comment:
This is the meat of the fix. Without this change, the print statement below prints a *different value* every time you hit the snapshot (I tested it). However, with the fix the value is always consistent. I believe this will fix our checksum issue, but I'm syncing a couple nodes right now to do a "final" test.